### ISSUE TITLE
Recipe for facets

### ISSUE THREAD

**Issue by mkborregaard** [JuliaPlots/Plots.jl#839]
It should be possible, and fairly easy, to make a userplot recipe for facet-grids, similar to subplot_grid in Gadfly, or facetgrid in ggplot2.
<img width="760" alt="skaermbillede 2017-05-17 kl 23 03 18" src="https://cloud.githubusercontent.com/assets/8429802/26176042/0d268a50-3b55-11e7-8d67-7cd80be6656c.png">

**Comment by mkborregaard** [comment]
The tricky thing would be to have the color bar align for all sublots.

**Comment by ValdarT** [comment]
Perhaps the most straight-forward (albeit not efficient) approach would be to reshape the data into a 'wide' format matrix? It should be possible to borrow code from DataFrames for the reshaping.

An ugly example (using DataFrames) to illustrate what I mean:
```
using Plots, RDatasets

data = dataset("datasets", "OrchardSprays")

cols = Symbol.(unique(data[:Treatment]))
widedata = unstack(data, :Treatment, :RowPos)

function preprocess!(widedata)
    tempcol = Vector{Float64}(size(widedata,1))
    for c in cols
        for r in 1:size(widedata,1)
            tempcol[r] = isna(widedata[r, c]) ? NaN : Float64(widedata[r, c])
        end
        widedata[c] = copy(tempcol)
    end
    widedata
end

preprocess!(widedata);

scatter(widedata[:ColPos], Array(widedata[cols]), layout=length(cols), legend=false, 
    color=widedata[:Decrease], title=reshape(cols,1,length(cols)))
```

![facetexample](https://user-images.githubusercontent.com/7242935/30126262-24e212f8-9344-11e7-8cba-c5dbeaba8640.png)

From there it would just be prettying the plot.

I know it's not perfect but perhaps this approach would be good enough.

BTW, I think that calls with numeric argument to color (e.g., `scatter(rand(20), color=rand(20))`) should add a colorscale legend automatically.

**Comment by mkborregaard** [comment]
"Prettying the plot" is what we do - I can tell you it's not "just" :-)

With regards to the plot you're overcomplicating it.
```julia
using StatPlots, RDatasets
data = dataset("datasets", "OrchardSprays")
@df data scatter(:ColPos, :RowPos, marker_z = :Decrease, group = :Treatment, layout = 8, label = "")
```
<img width="682" alt="skaermbillede 2017-09-06 kl 20 41 06" src="https://user-images.githubusercontent.com/8429802/30128709-ecbf7514-9343-11e7-9e96-49429b583398.png">

The difficult part is to make sure there's only one colorbar shown, and that that's shown outside all of the plots, with all the colors aligned to it. We also need the axis to merge so ticks are only shown at the edges.

**Comment by mkborregaard** [comment]
Oh and you don't pass the values you mean to control colors to `color` - that's Gadfly syntax. You pass them to `marker_z`. That automatically does show a color bar. `color` is for choosing the specific color scale.

**Comment by ValdarT** [comment]
Right. I obviously totally missed the point here. Thank you for the clarification!
I need to take the time to get to know Plots better. It is a really magical package.

**Comment by ValdarT** [comment]
Although, for what it's worth, for unifying the color bar the reshaping does work. : )

(Plotted the 'widedata' from pervious code snippet.)

![facetexample](https://user-images.githubusercontent.com/7242935/30130081-b93fb886-9350-11e7-9f47-eee8abf33515.png)

**Comment by mkborregaard** [comment]
You can also do it by adding `clim = extrema(:Decrease)...` to my call above.

**Comment by mkborregaard** [comment]
```julia
titles = reshape(sort(unique(data[:Treatment])), 1, :)
@df data scatter(:ColPos, :RowPos, marker_z = :Decrease, group = :Treatment, layout = 8, label = "", title = titles, clim = (extrema(:Decrease)...))
```
